libnetwork/drivers/bridge/port_mapping_linux.go

```diff
@@ -275,24 +276,96 @@ func (n *bridgeNetwork) setPerPortIptables(b *portBinding, enable bool) error {
 		return nil
 	}
 
-	natChain, _, _, _, err := n.getDriverChains(v)
-	if err != nil || natChain == nil {
+	if enabled, err := n.iptablesEnabled(v); err != nil || !enabled {
 		// Nothing to do, iptables/ip6tables is not enabled.
 		return nil
 	}
-	action := iptables.Delete
-	if enable {
-		action = iptables.Insert
-	}
-	return natChain.Forward(
-		action,
-		b.HostIP,
-		int(b.HostPort),
-		b.Proto.String(),
-		b.IP.String(),
-		int(b.Port),
-		n.getNetworkBridgeName(),
-	)
+
+	bridgeName := n.getNetworkBridgeName()
+	proxyPath := n.userlandProxyPath()
+	if err := setPerPortNAT(b, v, proxyPath, bridgeName, enable); err != nil {
+		return err
+	}
+	if err := setPerPortForwarding(b, v, bridgeName, enable); err != nil {
+		return err
+	}
+	return nil
+}
+
+func setPerPortNAT(b *portBinding, ipv iptables.IPVersion, proxyPath string, bridgeName string, enable bool) error {
```

Comment:
The issue with the way you've divided up your commits is that it's not clear what's new code and what's moved from one package to another. This makes review harder, but more importantly, future selves will have a hard time doing archeology.

Reply:
Not much has moved. Do you have any concrete suggestions?

Reply:
Forgot to note here - but... combined the delete-commits with the others. (I kept them separate to start with, in case there were any concerns about maintaining the old exported package interfaces.)